### ISSUE TITLE
Pairing thread devices failed

### ISSUE THREAD

**Issue by Leonater** [home-assistant-libs/python-matter-server#296]
I attempted to pair a eve energy smart plug. The plug is currently paired with HomeKit, I copied the setup code from the Apple Home app, then attempted to pair with the Home Assistant iOS app. The pairing only fails with thread, WiFi devices work as expected.

python-matter-server 3.4.0
iOS 16.4.1(a)

Matter server logs:
```
2023-05-22 18:59:33 raspberrypi matter_server.server.helpers.paa_certificates[1] INFO Fetching the latest PAA root certificates from DCL.
2023-05-22 18:59:34 raspberrypi matter_server.server.helpers.paa_certificates[1] INFO Fetched 0 PAA root certificates from DCL.
2023-05-22 18:59:34 raspberrypi matter_server.server.helpers.paa_certificates[1] INFO Fetching the latest PAA root certificates from Git.
2023-05-22 18:59:34 raspberrypi matter_server.server.helpers.paa_certificates[1] INFO Fetched 0 PAA root certificates from Git.
/usr/local/lib/python3.10/asyncio/events.py:73: RuntimeWarning: coroutine 'StorageController.async_save' was never awaited
  self._args = None
RuntimeWarning: Enable tracemalloc to get the object allocation traceback
2023-05-22 18:59:34 raspberrypi chip.CTL[1] INFO Setting attestation nonce to random value
2023-05-22 18:59:34 raspberrypi chip.CTL[1] INFO Setting CSR nonce to random value
2023-05-22 18:59:34 raspberrypi chip.CTL[1] INFO Starting commissioning discovery over BLE
2023-05-22 18:59:34 raspberrypi chip.CTL[1] INFO Starting commissioning discovery over DNS-SD
2023-05-22 18:59:34 raspberrypi chip.BLE[1] INFO BLE removing known devices.
2023-05-22 18:59:34 raspberrypi chip.BLE[1] INFO BLE initiating scan.
2023-05-22 18:59:34 raspberrypi chip.DL[1] ERROR Long dispatch time: 623 ms, for event type 2
2023-05-22 18:59:35 raspberrypi chip.CTL[1] INFO Discovered device to be commissioned over DNS-SD
2023-05-22 18:59:35 raspberrypi chip.CTL[1] INFO Attempting PASE connection to UDP:[replaced]:5540
2023-05-22 18:59:35 raspberrypi chip.EM[1] INFO <<< [E:36348i S:0 M:94836065] (U) Msg TX to 0:0000000000000000 [0000] --- Type 0000:20 (SecureChannel:PBKDFParamRequest)
2023-05-22 18:59:35 raspberrypi chip.IN[1] INFO (U) Sending msg 94836065 to IP address 'UDP:[replaced]:5540'
2023-05-22 18:59:35 raspberrypi chip.CTL[1] INFO Commission called for node ID 0x0000000000000001
2023-05-22 18:59:36 raspberrypi chip.IN[1] INFO (U) Sending msg 94836065 to IP address 'UDP:[replaced]:5540'
2023-05-22 18:59:36 raspberrypi chip.CTL[1] INFO Discovered device to be commissioned over DNS-SD
2023-05-22 18:59:37 raspberrypi chip.IN[1] INFO (U) Sending msg 94836065 to IP address 'UDP:[replaced]:5540'
2023-05-22 18:59:38 raspberrypi chip.CTL[1] INFO Discovered device to be commissioned over DNS-SD
2023-05-22 18:59:38 raspberrypi chip.IN[1] INFO (U) Sending msg 94836065 to IP address 'UDP:[replaced]:5540'
2023-05-22 18:59:41 raspberrypi chip.IN[1] INFO (U) Sending msg 94836065 to IP address 'UDP:[replaced]:5540'
2023-05-22 18:59:42 raspberrypi chip.CTL[1] INFO Discovered device to be commissioned over DNS-SD
2023-05-22 18:59:44 raspberrypi chip.BLE[1] ERROR BLE scan error: src/platform/Linux/bluez/ChipDeviceScanner.cpp:154: CHIP Error 0x00000032: Timeout
2023-05-22 18:59:44 raspberrypi chip.BLE[1] INFO Scan complete. No matching device found.
2023-05-22 18:59:45 raspberrypi chip.EM[1] ERROR Failed to Send CHIP MessageCounter:94836065 on exchange 36348i sendCount: 4 max retries: 4
2023-05-22 18:59:48 raspberrypi chip.SC[1] ERROR PASESession timed out while waiting for a response from the peer. Expected message type was 33
2023-05-22 18:59:48 raspberrypi chip.CTL[1] INFO Ignoring SecurePairingFailed status for now; we have more discovered devices to try
2023-05-22 18:59:48 raspberrypi chip.CTL[1] INFO Attempting PASE connection to UDP:[replaced]:5540
2023-05-22 18:59:48 raspberrypi chip.EM[1] INFO <<< [E:36349i S:0 M:94836066] (U) Msg TX to 0:0000000000000000 [0000] --- Type 0000:20 (SecureChannel:PBKDFParamRequest)
2023-05-22 18:59:48 raspberrypi chip.IN[1] INFO (U) Sending msg 94836066 to IP address 'UDP:[replaced]:5540'
2023-05-22 18:59:48 raspberrypi chip.CTL[1] INFO Commission called for node ID 0x0000000000000001
2023-05-22 18:59:48 raspberrypi chip.CTL[1] INFO Trying connection to commissionee over different transport
2023-05-22 18:59:49 raspberrypi chip.IN[1] INFO (U) Sending msg 94836066 to IP address 'UDP:[replaced]:5540'
2023-05-22 18:59:50 raspberrypi chip.IN[1] INFO (U) Sending msg 94836066 to IP address 'UDP:[replaced]:5540'
2023-05-22 18:59:52 raspberrypi chip.IN[1] INFO (U) Sending msg 94836066 to IP address 'UDP:[replaced]:5540'
2023-05-22 18:59:54 raspberrypi chip.IN[1] INFO (U) Sending msg 94836066 to IP address 'UDP:[replaced]:5540'
2023-05-22 18:59:58 raspberrypi chip.EM[1] ERROR Failed to Send CHIP MessageCounter:94836066 on exchange 36349i sendCount: 4 max retries: 4
2023-05-22 19:00:01 raspberrypi chip.SC[1] ERROR PASESession timed out while waiting for a response from the peer. Expected message type was 33
2023-05-22 19:00:01 raspberrypi chip.CTL[1] INFO Ignoring SecurePairingFailed status for now; we have more discovered devices to try
2023-05-22 19:00:01 raspberrypi chip.CTL[1] INFO Attempting PASE connection to UDP:[replaced]:5540
2023-05-22 19:00:01 raspberrypi chip.EM[1] INFO <<< [E:36350i S:0 M:94836067] (U) Msg TX to 0:0000000000000000 [0000] --- Type 0000:20 (SecureChannel:PBKDFParamRequest)
2023-05-22 19:00:01 raspberrypi chip.IN[1] INFO (U) Sending msg 94836067 to IP address 'UDP:[replaced]:5540'
2023-05-22 19:00:01 raspberrypi chip.CTL[1] INFO Commission called for node ID 0x0000000000000001
2023-05-22 19:00:01 raspberrypi chip.CTL[1] INFO Trying connection to commissionee over different transport
2023-05-22 19:00:02 raspberrypi chip.IN[1] INFO (U) Sending msg 94836067 to IP address 'UDP:[replaced]:5540'
2023-05-22 19:00:03 raspberrypi chip.IN[1] INFO (U) Sending msg 94836067 to IP address 'UDP:[replaced]:5540'
2023-05-22 19:00:04 raspberrypi chip.CTL[1] ERROR Discovery timed out
2023-05-22 19:00:04 raspberrypi chip.BLE[1] ERROR BleConnectionDelegate::CancelConnection is not implemented.
2023-05-22 19:00:04 raspberrypi chip.-[1] ERROR src/platform/Linux/BLEManagerImpl.cpp:732: CHIP Error 0x0000002D: Not Implemented at src/controller/SetUpCodePairer.cpp:551
2023-05-22 19:00:05 raspberrypi chip.IN[1] INFO (U) Sending msg 94836067 to IP address 'UDP:[replaced]:5540'
2023-05-22 19:00:08 raspberrypi chip.IN[1] INFO (U) Sending msg 94836067 to IP address 'UDP:[replaced]:5540'
2023-05-22 19:00:12 raspberrypi chip.EM[1] ERROR Failed to Send CHIP MessageCounter:94836067 on exchange 36350i sendCount: 4 max retries: 4
2023-05-22 19:00:14 raspberrypi chip.SC[1] ERROR PASESession timed out while waiting for a response from the peer. Expected message type was 33
2023-05-22 19:00:14 raspberrypi chip.CTL[1] INFO Ignoring SecurePairingFailed status for now; we have more discovered devices to try
2023-05-22 19:00:14 raspberrypi chip.CTL[1] INFO Attempting PASE connection to UDP:[replaced]:5540
2023-05-22 19:00:14 raspberrypi chip.EM[1] INFO <<< [E:36351i S:0 M:94836068] (U) Msg TX to 0:0000000000000000 [0000] --- Type 0000:20 (SecureChannel:PBKDFParamRequest)
2023-05-22 19:00:14 raspberrypi chip.IN[1] INFO (U) Sending msg 94836068 to IP address 'UDP:[replaced]:5540'
2023-05-22 19:00:15 raspberrypi chip.CTL[1] INFO Commission called for node ID 0x0000000000000001
2023-05-22 19:00:15 raspberrypi chip.CTL[1] INFO Trying connection to commissionee over different transport
2023-05-22 19:00:16 raspberrypi chip.IN[1] INFO (U) Sending msg 94836068 to IP address 'UDP:[replaced]:5540'
2023-05-22 19:00:17 raspberrypi chip.IN[1] INFO (U) Sending msg 94836068 to IP address 'UDP:[replaced]:5540'
2023-05-22 19:00:18 raspberrypi chip.IN[1] INFO (U) Sending msg 94836068 to IP address 'UDP:[replaced]:5540'
2023-05-22 19:00:21 raspberrypi chip.IN[1] INFO (U) Sending msg 94836068 to IP address 'UDP:[replaced]:5540'
2023-05-22 19:00:25 raspberrypi chip.EM[1] ERROR Failed to Send CHIP MessageCounter:94836068 on exchange 36351i sendCount: 4 max retries: 4
2023-05-22 19:00:28 raspberrypi chip.SC[1] ERROR PASESession timed out while waiting for a response from the peer. Expected message type was 33
2023-05-22 19:00:28 raspberrypi chip.ZCL[1] ERROR Secure Pairing Failed
2023-05-22 19:00:28 raspberrypi matter_server.server.client_handler[1] ERROR [548225869344] Error handling message: CommandMessage(message_id='cf6ba4b65f3e41d686a3a4a30a51cd55', command='commission_with_code', args={'code': 'replaced'})
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/site-packages/matter_server/server/client_handler.py", line 188, in _run_handler
    result = await result
  File "/usr/local/lib/python3.10/site-packages/matter_server/server/device_controller.py", line 154, in commission_with_code
    raise NodeCommissionFailed(
matter_server.common.errors.NodeCommissionFailed: Commission with code failed for node 1
```

**Comment by ArturoGuerra** [comment]
is everything under the same VLAN? and have you tried pinging the device's ip from within HA?

**Comment by Leonater** [comment]
It's all under the same VLAN. Ping from HA/raspberry doesn't work. But if i ping from my MacBook in the same network i will reach the device.

**Comment by ArturoGuerra** [comment]
do you have a ipv6 enabled on your network and if you do how are you acquiring an ipv6 address on HA?

**Comment by marcelveldt** [comment]
Is your HAOS running virtualized perhaps ? In that case somebody reported that enabling promiscuous might fix it. If ping(6) is not working, communicating with the matter devices will not work.

**Comment by Leonater** [comment]
> Is your HAOS running virtualized perhaps ? In that case somebody reported that enabling promiscuous might fix it. If ping(6) is not working, communicating with the matter devices will not work.

Yes it's running in docker. Does i need something like "ip link set eth0 promisc on" or docker MACVLAN?

**Comment by marcelveldt** [comment]
Running the matter server yourself (so without HAOS) is officially not supported but yeah you could give that a try, as well as making sure you have latest kernel patches etc. for the routing issues in the network manager.

**Comment by Leonater** [comment]
Thanks for the tip! I got it working. Apparently raspbian does not create the ipv6 route correctly.

`sudo ip route add *1::/64 via *2 dev eth0`

Replace: 
*1: thread ipv6 network address (see matter-server log)
*2: unique local address from border router

**Comment by wozz** [comment]
I'm having the same problem and was able to resolve by adding a route to the border router - however this solution seems fragile, what happens when the network topology changes? How are these routes normally configured automatically? Is there some component missing?

**Comment by Leonater** [comment]
I found an even better solution.
Check if "accept router advertisements" is on (1 or 2 in case IP forwarding is enabled):
`sysctl -n net.ipv6.conf.eth0.accept_ra`

If not, append below commands to `/etc/sysctl.conf`

```
net.ipv6.conf.eth0.accept_ra=1
net.ipv6.conf.eth0.accept_ra_rt_info_max_plen = 64
```

And add `noipv6` to `/etc/dhcpcd.conf`

**Note**: Remember to replace eth0 with your specific Wi-Fi/Ethernet interface.

@marcelveldt
Can you please add this requirement to the documentation?

**Comment by marcelveldt** [comment]
@Leonater nice find! Thanks for finding this all out, I think this will be appreciated by other users that want to run the docker image theirselves.

**Comment by wozz** [comment]
I think I found a different way to get it working - I tried updating the sysctl config, but it wasn't working for me. I updated the systemd-networkd config with a file in `/etc/systemd/network/01-accept-ra.network`

```
[Match]
Name=eth*

[Network]
DHCP=yes
IPv6AcceptRA=true
```

then `systemctl enable systemd-networkd`

I made various other changes while debugging this, so I guess to validate I would have to start over from scratch and see if it works on a clean OS install.

My route table now has routes to what I assume are all my thread border routers for the thread network subnet.

**Comment by ArturoGuerra** [comment]
yeah, that basically does the same as the previous one, most distros have it enabled by default as long as you have ipv6 enabled.

@marcelveldt we should probably add something to the readme saying something along the lines that RA should be enabled on the matter interface

**Comment by juliodee** [comment]
> I found an even better solution. Check if "accept router advertisements" is on (1 or 2 in case IP forwarding is enabled): `sysctl -n net.ipv6.conf.eth0.accept_ra`
> 
> If not, append below commands to `/etc/sysctl.conf`
> 
> ```
> net.ipv6.conf.eth0.accept_ra=1
> net.ipv6.conf.eth0.accept_ra_rt_info_max_plen = 64
> ```
> 
> And add `noipv6` to `/etc/dhcpcd.conf`
> 
> **Note**: Remember to replace eth0 with your specific Wi-Fi/Ethernet interface.
> 
> @marcelveldt Can you please add this requirement to the documentation?

I have been pulling my hair out for over 3 days trying to get matter over thread working with a Nanoleaf Essentials downlight. I have researched and tried everything I can. Came across this GitHub topic.

My environment is a QNAP, HA container and Python-Matter-Server container - both on network "host". I used both Container Station and Portainer, hoping the latter provided me more options. To the best of my knowledge, ipv6 is working (ie. I can see an FE80 address in HA Settings/System/Network/Network Adapter).

The only setting I don't have is --security-opt apparmor=unconfined because I don't know how to do this within the Portainer GUI. I also don't have the D-Bus parameter as it appears for Bluetooth commissioning and I don't think I would require this.

I have an Echo 4th Gen thread network and a Samsung Smartthings thread network (via a Q990C). I can connect to both via their respective apps, the Nanoleaf tends to prefer the Echo and also connects to the device whilst on the thread network, so everything works - except for Home Assistant being able to connect to the downlight.

- Trying the suggestions in this post, I can't find /etc/sysctl.conf exists.
- However, a sysctl -n net.ipv6.conf.eth1.accept_ra returns 1, so I assume the configuration file is "somewhere". 
- Trying sysctl -n net.ipv6.conf.eth0.accept_ra_rt_info_max_plen returns "unknown key". 
- I don't have /etc/dhcpcd.conf either. 
- In short, I have no idea where to insert these settings.

I don't know if there's a better error log, but the only errors I have been working off is:
![image](https://github.com/user-attachments/assets/38efac15-613c-4f62-972e-5a7f4e13ffe6)

If anyone can help how I can try these settings on a QNAP environment, would be appreciated!
